flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/InputGate.java

```diff
@@ -107,6 +107,20 @@
 		return availabilityHelper.getAvailableFuture();
 	}
 
+	/**
+	 * Returns the index of this gate in respect to the consuming operator instance.
+	 */
+	public int getGateIndex() {
+		throw new UnsupportedOperationException();
```

Comment:
Are there any benefits for implementing unsupported way in default?
If all the normal subclasses should implement it, it is better to define as abstract, otherwise the new added subclasses would not be aware of this method and cause unexpected exceptions in practice.

Reply:
I guess we can remove it from `InputGate` and only implement it inside `SingleInputGate` because all the usages currently are referring to the `SingleInputGate`, e.g. in `UnionInputGate` arguments and `InputChannel` component.

Reply:
It was a bit harder to do because of `InputGateWithMetrics` which also needs to have the `getIndex` as it is usually passed to `UnionInputGate`. So I added an extra commit that introduces `IndexedInputGate`. Only this `InputGate` has an `getIndex` and union gate can only be composed by these gates (turning the runtime into a compile time check).